### DISCUSSION TITLE
Lost all annotations

### DISCUSSION THREAD

**Discussion by ledg555** [sumatrapdfreader/sumatrapdf#3935] [Q&A]
Hi. I just wanted to ask: is there some shortcut to restore an original document or to suddenly delete all annotations? I was trying to save an annotation to the current document with the usual Ctrl+Shift+s but instead of s accidentally hit other key (I think w or e) and all my hundreds of annotations in a 700 pages book suddenly were gone.

**Comment by GitHubRulesOK** [comment]
Generally annotations are added to end of file and deletions are added to end of file.
So file will have both, and a session fixer can roll back to the previous state, as long as any editor has not been used. 
However the best tool for the job was abandoned some time ago as it was open source free ware with little commercial support. 

It should be "relatively  easy" to do it manually with any hex editor like "docking" a lambs tail, but you need to know where to cut the tail and why.

**Reply by WestCorkPalaeo** [reply]
Hi. I edited a file and inserted numerous annotations and comments. I saved regularly and while this was in action, all annotations were there, and remained accessible and editable (I did go back and add to a couple). When finished I saved the annotations and then closed the file.
Now, opening the file, no annotations are visible.
What can I do - if anything - to get them back?
I note a previous response suggested some hex editor but gave no hint as to how to 'dock' the lamb's tail.
I am happy to give this a go - I HAVE top give it a go. The file is a book that is to be sent for publication. Either I retrieve the annotations or I read through the whole book again. Please, any advice would be very useful.

**Reply by GitHubRulesOK** [reply]
dont edit in notepad but look at end of file the trailer should be visible something like this very small file actually a non functional file due to javascripting (no annotations with a `startxref 3089 %%EOF` (bigger the file bigger the number which is the decimal address needed for hexedit  or query
```
%PDF-1.7
%âãÏÓ
1 0 obj<</OpenAction 3 0 R/Pages 1 0 R/Type /Page>>endobj
3 0 obj<</Type /Action/S /JavaScript/JS (app.alert('XSS');)>> 
4 0 obj<</Filter /FlateDecode/First 4/Length 73/N 1/Type /ObjStm>>
stream
xÚ3V0à²±áÒ÷
VÐH,(ÐKÌI-*‰ÑPVÑ´ÖäÒVÐ÷J,KN.Ê,(áÒ©,HUÐwL.ÉÌÏã²³ã >MN
endstream
endobj
5 0 obj<</DecodeParms <</Columns 4/Predictor 12>>/Filter /FlateDecode/Length 37/Root 1 0 R/Size 6/Type /XRef/W [1 2 1]
>>
stream
xÚcb``øÏÄÈÀÏÈÄÀÀ d-e`úÏÈÀÄÀ¸† 6¸+
endstream
endobj
startxref
3089
%%EOF
```

Common problems are  
- %%EOF is missing and then the file can be unstable
- number is incorrect and thus in that case some additions  may be missing
- startxref is missing then its likely the file will be altered by some applications as missing the navigation pointer (so dont open and save in acrobat or others as the whole file  will be restructured possibly by deleting many objects)

The problem is finding a small sample so here is a file with annotation
```
%PDF-1.7
%ÂµÂ¶
1 0 obj <</Type/Catalog/Pages 2 0 R>> endobj
2 0 obj <</Type/Pages/Count 1/Kids [ 3 0 R ]>> endobj
3 0 obj <</Type/Page/MediaBox [ 0 0 72 72 ]/Rotate 0/Resources 4 0 R/Contents 6 0 R/Parent 2 0 R>> endobj
4 0 obj <</XObject<</Img3 7 0 R>>>> endobj
5 0 obj <</Length 4/Type/XObject/Subtype/Image/Width 3/Height 3/BitsPerComponent 1/ColorSpace/DeviceGray>>
stream
ÿÿÿ
endstream
endobj
6 0 obj <</Length 46>>
stream
1 0 0 -1 -0 72 cm
72 0 0 -72 0 72 cm
/Img3 Do
endstream
endobj

7 0 obj<</Length 27/Type/XObject/Subtype/Image/Width 3/Height 3/BitsPerComponent 8/SMask 5 0 R/ColorSpace/DeviceRGB>>
stream
ÿ   ÿ   ÿ ÿÿÿ ÿÿÿ    ÿÿÿ   
endstream
endobj

xref
0 8
0000000000 00002 f 
0000000015 00000 n 
0000000060 00000 n 
0000000114 00000 n 
0000000220 00000 n 
0000000263 00000 n 
0000000399 00000 n 
0000000493 00000 n 

trailer
<</Size 8/Root 1 0 R>>
startxref
664
%%EOF

3 0 obj
<</Type/Page/MediaBox[0 0 72 72]/Rotate 0/Resources 4 0 R/Contents 6 0 R/Parent 2 0 R/Annots[8 0 R 9 0 R]>>
endobj

8 0 obj
<</Type/Annot/Subtype/Text/Rect[12.264263 20.79234 30.264264 38.79234]/C[1 1 0]/Popup 9 0 R/P 3 0 R/F 28/M(D:20231212140546Z)/T(matin)/AP<</N 10 0 R>>/Contents(Hello World)>>
endobj

9 0 obj
<</Type/Annot/Subtype/Popup/Parent 8 0 R/Rect[32 -40 232 60]>>
endobj

10 0 obj
<</Type/XObject/Subtype/Form/BBox[0 0 16 16]/Matrix[1 0 0 1 0 0]/Length 72/Filter/FlateDecode>>
stream
xœ3T0T0P(Jç2T(ç2Ð3U aC0*JåJâÉ(è*˜()$çr(¤s„,€*ŒLÓÆLã Á\l
endstream
endobj

xref
3 1
0000000886 00000 n 
8 3
0000001010 00000 n 
0000001201 00000 n 
0000001280 00000 n 

trailer
<</Size 11/Root 1 0 R/Prev 664>>
startxref
664
%%EOF
```

Now you should see there are 2 endings the original `startxref 664 %%EOF` and the annotation terminates with  `startxref 664 %%EOF` (THAT SEEMS TO BE THE CAUSE OF ERROR)

Beware even a source file can often have multiple endings, so you work backwards from end searching and looking for %%EOF to find the best one, but you can do that one by one and stop when file save looks "good"

In example above you would only be able to fix the file ending or cut tail off in hex editor to the original as there are no other %%EOF

Typical to the annotations are existing objects modified hence the 3 0 obj is a modified duplicate of the same source entry with pointers to tha annotation which is now `8 0 obj` and so what are 9 and 10 ? they are the obligatory PDF bloat to provide eye candy such as a compressed icon at 10 and a placement at 9 !

**Comment by GitHubRulesOK** [comment]
@kjk 
in that above description **at the END I saw there was an error** the 664 was duplicate but should be like this MuPDF incremental example since the file is more than twice as big (later edit I am not sure where above example came from as its no longer around !)
```
%PDF-1.7
%ÂµÂ¶
1 0 obj <</Type/Catalog/Pages 2 0 R>> endobj
2 0 obj <</Type/Pages/Count 1/Kids [ 3 0 R ]>> endobj
3 0 obj <</Type/Page/MediaBox [ 0 0 72 72 ]/Rotate 0/Resources 4 0 R/Contents 6 0 R/Parent 2 0 R>> endobj
4 0 obj <</XObject<</Img3 7 0 R>>>> endobj
5 0 obj <</Length 4/Type/XObject/Subtype/Image/Width 3/Height 3/BitsPerComponent 1/ColorSpace/DeviceGray>>
stream
ÿÿÿ
endstream
endobj
6 0 obj <</Length 46>>
stream
1 0 0 -1 -0 72 cm
72 0 0 -72 0 72 cm
/Img3 Do
endstream
endobj

7 0 obj<</Length 27/Type/XObject/Subtype/Image/Width 3/Height 3/BitsPerComponent 8/SMask 5 0 R/ColorSpace/DeviceRGB>>
stream
ÿ   ÿ   ÿ ÿÿÿ ÿÿÿ    ÿÿÿ   
endstream
endobj

xref
0 8
0000000000 00002 f 
0000000015 00000 n 
0000000060 00000 n 
0000000114 00000 n 
0000000220 00000 n 
0000000263 00000 n 
0000000399 00000 n 
0000000493 00000 n 

trailer
<</Size 8/Root 1 0 R>>
startxref
664
%%EOF

3 0 obj
<</Type/Page/MediaBox[0 0 72 72]/Rotate 0/Resources 4 0 R/Contents 6 0 R/Parent 2 0 R/Annots[8 0 R 9 0 R]>>
endobj

8 0 obj
<</Type/Annot/Subtype/Text/Rect[10.75 14 28.75 32]/C[1 1 0]/Popup 9 0 R/P 3 0 R/F 28/M(D:20231212142832Z)/T(lez)/AP<</N 10 0 R>>/Contents(hello world)>>
endobj

9 0 obj
<</Type/Annot/Subtype/Popup/Parent 8 0 R/Rect[32 -40 232 60]>>
endobj

10 0 obj
<</Type/XObject/Subtype/Form/BBox[0 0 16 16]/Matrix[1 0 0 1 0 0]/Length 99>>
stream
1 1 0 rg
1 w
0.5 0.5 15 15 re
b
1 0 0 -1 4 12 cm
0 g
0 0 8 1 re
0 2 8 1 re
0 4 8 1 re
0 6 8 1 re
f

endstream
endobj

xref
3 1
0000000886 00000 n 
8 3
0000001010 00000 n 
0000001179 00000 n 
0000001258 00000 n 

trailer
<</Size 11/Root 1 0 R/Prev 664>>
startxref
1469
%%EOF
```

**Reply by WestCorkPalaeo** [reply]
OK, great, thanks.

You say DONT edit in notepad.

can you recommend a hex editor?
is notepad++ any good for this?

**Reply by WestCorkPalaeo** [reply]
Ah, OK, not to worry. I have a hexeditor. The annotations are still in the file - thank god! Now to find out how to make them display

**Reply by GitHubRulesOK** [reply]
@WestCorkPalaeo 
the problem is the null character i.e. nothing so most text editors will either remove or replace nothing with space  :-)
that is an ability that allows me to use charcoal as blacking in MS notepad but corrupts any fonts and once lengths drift a PDF will have all the wrong addresses throughout.

HxD has some nice abilities
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/0c476c71-bf83-4f67-9def-dd2f01656870)

**Reply by WestCorkPalaeo** [reply]
OK. I have found that if I delete from the last EOF back to the previous EOF then I get a bunch of annotations - but not all of them. It looks as if they are in 'lumps' - possibly the lumps as I saved them periodically.
So I can try deleting from the last EOF to the previous one, then save the file with an incremental number, and keep doing this until I have all the annotations back in a host of files. It at least will mean I dont have to do it all again.
The big question is, why would this have happened? I haven't had this problem before.

**Reply by GitHubRulesOK** [reply]
i cant replicate the problem I show above but if the startxref is not itself incremented the last group will "disappear" so the fix is simply update that last entry to start the last xref as shown

**Reply by GitHubRulesOK** [reply]
@WestCorkPalaeo 

from my experience above there is a possibility that  all annotations are stored but not seen when that last startxref is a duplicate of an earlier one. (that should not happen and I cannot repeat it happening, but take that example at face value. If I open the file in HXD and see that the last xref is not correct I can try to change it.
wrong
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/a2306a90-3565-4a38-a817-2e587e699945)
fixed
**note due to an extra byte in that number from 3 digits to 4 the the EOF also needs to move on one extra byte.**
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/fa7a1383-9786-47e6-b59f-bfd4dc0bd8de)

**Comment by WestCorkPalaeo** [comment]
Well, I'm all fixed now thanks for your help.
I have to add that this confirms my preference for Sumatra as the pdf reader - the fact that I can recover data that seemed lost is a big deal.
Thanks again.

**Reply by qwest0** [reply]
hello i was typing a long annotation but when typing i pressed delete button and it deleted the whole annotation , is it possible to undo this or find the typed annotation , this is happening repeatedly as its a habit of pressing del button while typing something wrong.

**Reply by GitHubRulesOK** [reply]
@qwest0 
There is no assurance of finding deleted annotations if not saved first
it is a fine balance between saving on a regular basis which appends bloating annotations (and thus still finding them after deletion)
 compared to lost time when deleted accidentally before save.
The only way to know is look in the tail of the file with a hex editor. 
Above losses due to wrong pointers is different to pressing delete

**Reply by qwest0** [reply]
thank you for replying sir , i tried with hex editor , i find only the last saved annotation typed , the one i was typing isnt there. should i delete in hex something to see where the typed annotation present or will it come back ? in what way i can post the hex pic here sir ?

**Reply by GitHubRulesOK** [reply]
If it is not in the file it is most likely NOT recoverable,
it is usually gone as PC memory frequently gets reused for other tasks.

Similar to when files are accidentally deleted,
 the flagged as free space in modern solid disk drive computers is often reallocated first,
 unlike older slow hard drives.

So recovery of lost data is nowadays less likely to succeed, than in the past.

**Reply by qwest0** [reply]
ok sir , thanks for explanations, sir kindly request you to pls consider change the del function while typing annotations to not del the annotation box itself . thank you.
